logbook-httpclient/src/main/java/org/zalando/logbook/httpclient/LocalRequest.java

```diff
@@ -16,21 +17,110 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Stream;
 
 import static java.nio.charset.StandardCharsets.UTF_8;
 import static java.util.stream.Collectors.groupingBy;
 import static java.util.stream.Collectors.mapping;
 import static java.util.stream.Collectors.toList;
 import static org.apache.http.util.EntityUtils.toByteArray;
+import static org.zalando.fauxpas.FauxPas.throwingUnaryOperator;
 
 final class LocalRequest implements org.zalando.logbook.HttpRequest {
 
+    private interface State {
+
+        default State with() {
+            return this;
+        }
+
+        default State without() {
+            return this;
+        }
+
+        default State buffer(final HttpRequest request) throws IOException {
+            return this;
+        }
+
+        default byte[] getBody() {
+            return new byte[0];
+        }
+
+    }
+
+    private static final class Unbuffered implements State {
+
+        @Override
+        public State with() {
+            return new Offering();
+        }
+
+    }
+
+    private static final class Offering implements State {
+
+        @Override
+        public State without() {
+            return new Unbuffered();
+        }
+
+        @Override
+        public State buffer(final HttpRequest request) throws IOException {
+            if (request instanceof HttpEntityEnclosingRequest) {
+                final HttpEntityEnclosingRequest original = (HttpEntityEnclosingRequest) request;
+                if (original.getEntity() == null) {
+                    return new Passing();
+                } else {
+                    final byte[] body = toByteArray(original.getEntity());
+                    original.setEntity(new ByteArrayEntity(body));
+                    return new Buffering(body);
+                }
+            } else {
+                return new Passing();
+            }
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static final class Buffering implements State {
+
+        private final byte[] body;
+
+        @Override
+        public State without() {
+            return new Ignoring(body);
+        }
+
+        @Override
+        public byte[] getBody() {
+            return body;
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static final class Ignoring implements State {
+
+        private final byte[] body;
+
+        @Override
+        public State with() {
+            return new Buffering(body);
+        }
+
+    }
+
+    private static final class Passing implements State {
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)